rest-api/dao/biobank_order_dao.py

```diff
@@ -333,10 +390,150 @@ def to_client_json(self, model):
+      resource.amendedInfo = self._to_handling_info(model.amendedUsername,
+                                                      model.amendedSiteId)
 
     self._add_identifiers_to_resource(resource, model)
     self._add_samples_to_resource(resource, model)
     client_json = resource.as_json()  # also validates required fields
     client_json['id'] = model.biobankOrderId
     del client_json['resourceType']
     return client_json
+
+  def _do_update(self, session, order, existing_obj):
+    order.lastModified = clock.CLOCK.now()
+    order.biobankOrderId = existing_obj.biobankOrderId
+    order.orderStatus = BiobankOrderStatus.AMENDED
+    order.amendedTime = clock.CLOCK.now()
+    order.logPosition = LogPosition()
+    order.version += 1
+    # Ensure that if an order was previously cancelled/restored those columns are removed.
+    self._clear_cancelled_and_restored_fields(order)
+
+    super(BiobankOrderDao, self)._do_update(session, order, existing_obj)
+    session.add(order.logPosition)
+
+    self._refresh_participant_summary(session, order)
+    self._update_history(session, order)
+    self._update_identifier_history(session, order)
+    self._update_sample_history(session, order)
+
+  def update_with_patch(self, id_, resource, expected_version):
+    """creates an atomic patch request on an object. It will fail if the object
+    doesn't exist already, or if obj.version does not match the version of the existing object.
+    May modify the passed in object."""
+    with self.session() as session:
+      obj = self.get_with_children_in_session(session, id_, for_update=True)
+      return self._do_update_with_patch(session, obj, resource, expected_version)
+
+  def _do_update_with_patch(self, session, order, resource, expected_version):
+    self._validate_patch_update(order, resource, expected_version)
+    order.lastModified = clock.CLOCK.now()
+    order.logPosition = LogPosition()
+    order.version += 1
+    if resource['status'].lower() == 'cancelled':
+      order.amendedReason = resource['amendedReason']
+      order.cancelledUsername = resource['cancelledInfo']['author']['value']
+      order.cancelledSiteId = get_site(resource['cancelledInfo'])
+      order.cancelledTime = clock.CLOCK.now()
+      order.orderStatus = BiobankOrderStatus.CANCELLED
+    elif resource['status'].lower() == 'restored':
+      order.amendedReason = resource['amendedReason']
+      order.restoredUsername = resource['restoredInfo']['author']['value']
+      order.restoredSiteId = get_site(resource['restoredInfo'])
+      order.restoredTime = clock.CLOCK.now()
+      order.orderStatus = BiobankOrderStatus.UNSET
+    else:
+      raise BadRequest('status must be restored or cancelled for patch request.')
+
+    self._update_history(session, order)
+    self._update_identifier_history(session, order)
+    self._update_sample_history(session, order)
+    super(BiobankOrderDao, self)._do_update(session, order, resource)
```

Comment:
I noticed that we do super._do_update() in _do_update() before update_history(), etc. and here we do it afterwards. Not sure if it matters...  but we might want to be consistent.